planner/cascades/transformation_rules.go

```diff
@@ -1670,3 +1672,112 @@ func pushLimitDownOuterJoinToChild(limit *plannercore.LogicalLimit, outerGroup *
 	newLimitGroup.SetChildren(outerGroup)
 	return memo.NewGroupWithSchema(newLimitGroup, outerGroup.Prop.Schema)
 }
+
+// EliminateOuterJoinBelowAggregation eliminate the outer join which below aggregation.
+type EliminateOuterJoinBelowAggregation struct {
+	baseRule
+}
+
+// NewRuleEliminateOuterJoinBelowAggregation creates a new Transformation EliminateOuterJoinBelowAggregation.
+// The pattern of this rule is `Aggregation->Join->X`.
+func NewRuleEliminateOuterJoinBelowAggregation() Transformation {
+	rule := &EliminateOuterJoinBelowAggregation{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandAggregation,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandJoin, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *EliminateOuterJoinBelowAggregation) Match(expr *memo.ExprIter) bool {
+	joinType := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalJoin).JoinType
+	return joinType == plannercore.LeftOuterJoin || joinType == plannercore.RightOuterJoin
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to eliminate outer join which below aggregation.
+func (r *EliminateOuterJoinBelowAggregation) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	agg := old.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	joinExpr := old.Children[0].GetExpr()
+	join := joinExpr.ExprNode.(*plannercore.LogicalJoin)
+
+	var innerChildIdx int
+	switch join.JoinType {
+	case plannercore.LeftOuterJoin:
+		innerChildIdx = 1
+	case plannercore.RightOuterJoin:
+		innerChildIdx = 0
+	default:
+		return nil, false, false, nil
+	}
+	outerGroup := joinExpr.Children[1^innerChildIdx]
+	innerGroup := joinExpr.Children[innerChildIdx]
+
+	outerUniqueIDs := set.NewInt64Set()
+	for _, outerCol := range outerGroup.Prop.Schema.Columns {
+		outerUniqueIDs.Insert(outerCol.UniqueID)
+	}
+	// only when agg only use the columns from outer table can eliminate outer join.
+	if !plannercore.IsColsAllFromOuterTable(agg.GetUsedCols(), outerUniqueIDs) {
+		return nil, false, false, nil
+	}
+	// outer join elimination with duplicate agnostic aggregate functions.
+	_, aggCols := plannercore.GetDupAgnosticAggCols(agg, nil)
+	if len(aggCols) > 0 {
+		newAggExpr := memo.NewGroupExpr(agg)
+		newAggExpr.SetChildren(outerGroup)
+		return []*memo.GroupExpr{newAggExpr}, true, false, nil
+	}
+	// outer join elimination without duplicate agnostic aggregate functions.
+	innerJoinKeys := join.ExtractJoinKeys(innerChildIdx)
+	contain, err := isInnerJoinKeysContainUniqueKey(innerGroup, innerJoinKeys)
```

Comment:
It seems we haven't add any test case which `innerJoinKeys contain UniqueKey`.

Reply:
integration_test?

Reply:
primary key is also uniqueKey?

Reply:
Yes